### PR TITLE
fix: Query parameter not being validated properly

### PR DESCRIPTION
**Pull request by KevLehman**
## Issue(s)

https://rocketchat.atlassian.net/browse/CORE-1787

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved query validation to avoid unnecessary checks and correctly handle objects with null prototypes.

* **Refactor**
  * Updated validation flow to consistently sanitize and safely process nested queries and removed-risk properties.

* **Tests**
  * Expanded record-detection unit tests to cover additional edge cases, wrapper objects, and array distinctions.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: apps/meteor/app/api/server/lib/isValidQuery.ts [hunk 1]
```diff
@@ -21,7 +24,7 @@ export const isValidQuery: {
 );
 
 const verifyQuery = (query: Query, allowedAttributes: string[], allowedOperations: string[], parent = ''): boolean => {
-	return Object.entries(removeDangerousProps(query)).every(([key, value]) => {
+	return Object.entries(removeDangerousProps({ ...query })).every(([key, value]) => {
```
Comment (KevLehman) [inline apps/meteor/app/api/server/lib/isValidQuery.ts:27]: Note for reviewers: since `query` was already `clean` at this point, it becomes an object with a `null prototype`.

When `removeDangerousProps` is called with it as param, then the call to `v.hasOwnProperty` fails (because the prototype is null)

File: packages/tools/src/isRecord.ts [hunk 2]
```diff
@@ -3,5 +3,11 @@ export function isRecord(record: unknown): record is Record<string | number | sy
 		return false;
 	}
 
-	return Object.getPrototypeOf(record).constructor === Object;
+	const prototype = Object.getPrototypeOf(record);
+
+	if (prototype === null) {
```
Comment (KevLehman) [inline packages/tools/src/isRecord.ts:12]: Note for reviewers: This would allow code like
```
const boxed = new String('hello');
console.log(typeof boxed);

Object.setPrototypeOf(boxed, null);
```

In that case, `boxed` would be a `null prototype` object. But, why should we write that code in the first place? 🤔 

Code like
```
const str = 'sadfasdfas'
Object.setPrototypeOf(str, null)
```

would not match cause the prototype doesn't changes on those cases

File: apps/meteor/app/api/server/lib/isValidQuery.ts
Comment (ricardogarim) [inline apps/meteor/app/api/server/lib/isValidQuery.ts]: Would it be necessary to add the `isRecord` check here as well, based on your [comment](https://github.com/RocketChat/Rocket.Chat/pull/38315/changes#r2722775239)?

Comment (KevLehman) [reply apps/meteor/app/api/server/lib/isValidQuery.ts]: Think it would hurt nobody

File: apps/meteor/app/api/server/helpers/parseJsonQuery.ts [hunk 3]
```diff
@@ -142,7 +142,8 @@ export async function parseJsonQuery(api: GenericRouteExecutionContext): Promise
 			}
 		}
 
-		if (queryFields && !isValidQuery(query, queryFields || ['*'], queryOperations ?? pathAllowConf.def)) {
+		const containsQueryFields = queryFields.length > 0;
+		if (containsQueryFields && !isValidQuery(query, containsQueryFields ? queryFields : ['*'], queryOperations ?? pathAllowConf.def)) {
```
[the remark this answers is not in the document]
Comment (ricardogarim) [reply apps/meteor/app/api/server/helpers/parseJsonQuery.ts]: true thing. if `containsQueryFields` is false, the second condition will not be tested. if so, you could alway pass `queryFields` to that instead of having the ternary.

File: apps/meteor/app/api/server/lib/isValidQuery.ts
Comment (ricardogarim) [reply apps/meteor/app/api/server/lib/isValidQuery.ts]: at this point we will have a null proto so think we should also use the `isRecord` check, right?

Comment (KevLehman) [inline apps/meteor/app/api/server/lib/isValidQuery.ts]:
```suggestion
			if (isRecord(value)) {
```

changed the other one in the file and left this :dum:

File: apps/meteor/app/api/server/helpers/parseJsonQuery.ts (hunk 3 above)
Comment (KevLehman) [inline apps/meteor/app/api/server/helpers/parseJsonQuery.ts]:
```suggestion
```

Comment (KevLehman) [inline apps/meteor/app/api/server/helpers/parseJsonQuery.ts]:
```suggestion
		if (containsQueryFields && !isValidQuery(query, queryFields, queryOperations ?? pathAllowConf.def)) {
```
